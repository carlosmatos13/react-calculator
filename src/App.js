import React, { useState } from 'react';
import Calculator from './Calculator/Calculator';
import './App.css';

const App = () => {
  // Set initial states for the values and operator
  const [initialValue, setValue] = useState('0');
  const [prevValue, setPrevValue] = useState(null);
  const [hasOperator, setHasOperator] = useState(false);
  const [operator, setOperator] = useState(null);

  // Formats Current value
  let formattedValue = parseFloat(initialValue).toLocaleString('en-US', {
    useGrouping: true,
    maximumFractionDigits: 5
  })

  //Object with functions to operate 
  const calculate = {
    '/': (prevValue, nextValue) => prevValue / nextValue,
    '*': (prevValue, nextValue) => prevValue * nextValue,
    '+': (prevValue, nextValue) => prevValue + nextValue,
    '-': (prevValue, nextValue) => prevValue - nextValue,
    '=': (nextValue) => nextValue
  }

  // Display current value on screen
  const displayValueHandler = (digit) => {
    if (hasOperator) {
      setValue(digit)
      setHasOperator(false)
    } else {
      setValue(
        (initialValue === "0") ? digit : initialValue + digit
      )
    }
  }
  // Clears display and reset states
  const clearDisplayHandler = () => {
    setValue("0");
    setPrevValue(null);
    setHasOperator(false);
    setOperator(null);
  }

  // This function handles decimal values 
  const addDotHandler = () => {
    if (!initialValue.includes('.')) {
      setValue(initialValue + '.')
      setHasOperator(false)
    }
  }
  // Calculate Percentage of Current Value
  const percentageHandler = () => {
    setValue(
      String(initialValue / 100)
    )
  }

  // Toggle sign of current value
  const toggleSignHandler = () => {
    if (initialValue !== '0') {
      setValue((initialValue.charAt(0) === '-')
        ? initialValue.substr(1)
        : '-' + initialValue
      )
    }
  }

  // This function takes current operator and previous value ands sets current value with result
  const operatorHandler = (nextOperator) => {
    const nextValue = parseFloat(initialValue)

    if (prevValue == null) {
      setPrevValue(nextValue)
    } else if (operator) {
      const result = calculate[operator](prevValue, nextValue);
      setPrevValue(result)
      setValue(String(result))
    }
    
    setHasOperator(true);
    setOperator(nextOperator);
  }

  return (
    <div className="App">
      <Calculator
        displayValue={displayValueHandler}
        clearDisplay={clearDisplayHandler}
        addDot={addDotHandler}
        percentage={percentageHandler}
        toggleSign={toggleSignHandler}
        operator={operatorHandler}
        display={formattedValue} />
    </div>
  );
}

export default App;
