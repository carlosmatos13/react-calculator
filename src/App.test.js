import React from 'react';
import renderer from 'react-test-renderer';
import App from './App';


describe('renders correctly', () => {
    it('matches the snapshot',() => {
        const tree = renderer.create(<App />).toJSON();
        expect(tree).toMatchSnapshot();
    });
})