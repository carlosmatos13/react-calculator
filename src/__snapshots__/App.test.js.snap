// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator component matches the snapshot 1`] = `
<div
  className="App"
>
  <div
    className="container"
  >
    <div
      id="calculator"
    >
      <div
        id="display"
      >
        <div
          className="scaletext-wrapper"
          style={
            Object {
              "fontSize": "inherit",
              "height": "100%",
              "overflow": "hidden",
              "width": "100%",
            }
          }
        >
          <p
            fontSize="inherit"
          >
            0
          </p>
        </div>
      </div>
      <div
        className="grid-container"
      >
        <button
          className="operator"
          onClick={[Function]}
        >
          AC
        </button>
        <button
          className="operator"
          onClick={[Function]}
        >
          ±
        </button>
        <button
          className="operator"
          onClick={[Function]}
        >
          %
        </button>
        <button
          className="operator right"
          onClick={[Function]}
        >
          ÷
        </button>
        <button
          className="number"
          onClick={[Function]}
        >
          7
        </button>
        <button
          className="number"
          onClick={[Function]}
        >
          8
        </button>
        <button
          className="number"
          onClick={[Function]}
        >
          9
        </button>
        <button
          className="operator right"
          onClick={[Function]}
        >
          x
        </button>
        <button
          className="number"
          onClick={[Function]}
        >
          4
        </button>
        <button
          className="number"
          onClick={[Function]}
        >
          5
        </button>
        <button
          className="number"
          onClick={[Function]}
        >
          6
        </button>
        <button
          className="operator right"
          onClick={[Function]}
        >
          -
        </button>
        <button
          className="number"
          onClick={[Function]}
        >
          1
        </button>
        <button
          className="number"
          onClick={[Function]}
        >
          2
        </button>
        <button
          className="number"
          onClick={[Function]}
        >
          3
        </button>
        <button
          className="operator right"
          onClick={[Function]}
        >
          +
        </button>
        <button
          className="number"
          id="numero0"
          onClick={[Function]}
        >
          0
        </button>
        <button
          className="number"
          onClick={[Function]}
        >
          .
        </button>
        <button
          className="operator right"
          onClick={[Function]}
        >
          =
        </button>
      </div>
    </div>
  </div>
</div>
`;
